tests/middleware/test_trace_logging.py

```diff
@@ -61,11 +62,13 @@ async def app(scope, receive, send):
 def test_trace_logging(capsys):
     class CustomServer(Server):
         def install_signal_handlers(self):
-            pass
+            self.signal_event = asyncio.Event()
+            self.shutdown_trigger = self.signal_event.wait  # type: ignore
```

Comment:
Super interesting — I ended up adding a similar `shutdown_trigger` idea in #863 (wording taken from Hypercorn's own `shutdown_trigger` option), because in some cases / tests I needed to be able to shutdown the server without having to send a request (`limit_max_requests` insufficient) and without having to send a signal.

This makes me think this idea is sort of the right spot: we need an event that can be set from outside saying "okay, please shutdown now" that's not tied to OS signals.

You can take a look at my `server.py` implementation, could be helpful: https://github.com/encode/uvicorn/blob/fm/async-agnostic/uvicorn/_async_agnostic/server.py